### ISSUE TITLE
Need a soonish fix for current usage of code generation

### ISSUE THREAD

**Issue by AlexanderWert** [open-telemetry/build-tools#242]
- In the past the way build-tools were used for code generation (in many language SDKs) assumed a strict separation between resource attributes and semantic attributes
- This assumption is not true anymore, as there are attributes that can be used as both, resource and semantic attributes
- Currently, there's no (easy/at all?) way to use the build-tools for code generation in a way that the above separation applies if we have namespaces in the registry that would contain both, resource and semantic attributes
- With the migration to a registry-based attributes model, the current limitation 

The above leads to the fact that some work on semantic conventions is blocked, because we cannot introduce namespaces with a mix of resource and semantic attributes.

**Comment by AlexanderWert** [comment]
\cc @lmolkova FYI: I created this issue to track all ongoing semconv PRs that are blocked by this issue.

**Comment by joaopgrassi** [comment]
Given https://github.com/open-telemetry/build-tools/pull/243, do we still need anything else to close this? @AlexanderWert

**Comment by lmolkova** [comment]
Adding a list of remaining things not yet addressed in #243

To merge the PR to main:
- [ ] File names should follow per-language pattern (`client_attrubutes.py`, `ClientAttrubutes.java`) 
- [ ] Update PRs in Java (https://github.com/open-telemetry/semantic-conventions-java/pull/41, https://github.com/open-telemetry/opentelemetry-python/pull/3586) and get approvals from Java and Python SIGs

Can be done as separate PRs:
- [ ] Add examples for attributes and metric generation (common case, split by stability)
- [ ] Improve docs 

Done, needs approvals:
- [ ] Make stability and deprecation independent properties #244

Needs discussion:
- [ ]  Codegen: proper casing for multiword attribute/metric names #599 - AFAIK it's not blocking. None of the affected attributes are stable **yet**

**Comment by jsuereth** [comment]
I created a prototype to help with filenames: https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/main/internal/resourcemapping/resourcemapping.go#L143

TL;DR; is we can allow `--output` filename to be a JINJA2 template itself and provide different versions of the "prefix" for the file to the template,e .g. camel-case, snake-case, ....

**Comment by lmolkova** [comment]
I believe we have just two items left, but I don't consider them blocking:
- don't create file if template filtered out all attributes - see [this](https://github.com/open-telemetry/semantic-conventions-java/blob/b6de6815e13e2fa4d80d5f0ef3e54f3441b3f8f2/buildscripts/templates/SemanticAttributes.java.j2#L35) for the details:

  - We have means to not create empty files, it's just not super-convenient. 
  - If we implemented it, it would not change generated code in any way, but implementation is not trivial.

- automate updating Schema URLs list (see [this on python](https://github.com/open-telemetry/opentelemetry-python/blob/9176adec3e62a0cd5f306715242e7d4829844965/opentelemetry-semantic-conventions/src/opentelemetry/semconv/schemas.py) and [this on java](https://github.com/open-telemetry/semantic-conventions-java/pull/41/files#diff-cdaff9d7dfb2d7006d135fc7f61d8435889574cd766d691e2ca052ccba7a1baf)):

  - This is something we can't achieve with Jinja templates. Also it's easy to do manually (add new constant to the list of schema urls when generating new semconv). Java also added check for it.

So the only thing before we start merging to main is getting green light from Java and Python - I'll check with Python tomorrow.

**Comment by lmolkova** [comment]
@jsuereth That's what I was also thinking. 

Currently filter is fully controlled by the template and python code does not know, so can't prevent empty files.
At the same time, we can make filters first-class citizens (e.g. stable, experimental). This would also save a tiny bit of effort in jinja templates.

If we did this, we could reuse existing `--only` cmd line which essentially applies filter  on signal. We could change it to accept  stability filters like `codegen --only metrics,stable` or `codegen --only experimental`.

With `codegen --only metrics,stable` SIGs would need to run codegen script for all combinations they want.
With `codegen --only experimental` SIG would likely have more filters (for signals) and they'd need to take care of empty files on their own.

I feel this feature (at least how I see it) brings diminishing returns or just needs more thought than I put into it.

**Comment by lmolkova** [comment]
I believe this can be closed - we got configuration from Java and Python SIGs that codegen works well. If there are any further improvements, we can track them in new issues. Thanks everyone!
